### PR TITLE
PipeReader.CopyToAsync(Stream) should not perform zero-length writes

### LINKED ISSUE #85733: PipeReader.CopyToAsync(Stream) sometimes does a zero length write
### Description

PipeReader.CopyToAsync(Stream) copies data from the pipe into the given Stream. However, under some circumstances it will call Stream.WriteAsync with a zero length buffer. This can be a problem if the Stream is only expecting a certain amount of data before automatically closing, such as an HTTP Content-Length response (HttpSys server), causing an ObjectDisposedException.

### Reproduction Steps

```c#
using System.Buffers;
using System.IO.Pipelines;

var pipe = new Pipe();
// await pipe.Writer.WriteAsync(new byte[100]); // Optional

var copyTask = pipe.Reader.CopyToAsync(new FakeStream());

pipe.Writer.Write(Array.Empty<byte>());
pipe.Writer.Complete();
await copyTask;

class FakeStream : Stream
{
    public override bool CanRead => false;

    public override bool CanSeek => false;

    public override bool CanWrite => true;

    public override long Length => throw new NotImplementedException();

    public override long Position { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

    public override void Flush() { }

    public override int Read(byte[] buffer, int offset, int count) => throw new NotImplementedException();

    public override long Seek(long offset, SeekOrigin origin) => throw new NotImplementedException();

    public override void SetLength(long value) => throw new NotImplementedException();

    public override void Write(byte[] buffer, int offset, int count) => throw new NotImplementedException();

    public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
    {
        Console.WriteLine("Writing: " + buffer.Length);
        return default;
    }
}
```

### Expected behavior

Stream.WriteAsync should not be called if the buffer length is 0.

### Actual behavior

Stream.WriteAsync may be called with a buffer length of 0.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 6, 7.

### Other information

https://github.com/dotnet/runtime/blob/8d35268c39cbc21c595265c51260518316032497/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs#L249-L251

TryGet is returning true with a zero length buffer.

### PR DESCRIPTION
**Pull request by mgravell**
Fix https://github.com/dotnet/runtime/issues/85733

Test for empty buffers in `CopyToAsyncCore` - do not invoke the write callback if empty

### CODE REVIEW

**Review by Tratcher — APPROVED** [review]

File: src/libraries/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs [hunk 1]
```diff
@@ -322,5 +322,60 @@ public async Task CopyToAsyncPipeWriterCopiesRemainderAfterReadingSome()
...
+            public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+            {
+                Check(count);
+                return base.BeginWrite(buffer, offset, count, callback, state);
+            }
+            public override void WriteByte(byte value)
+            {
+                Check(1);
+                base.WriteByte(value);
+            }
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs:378]: Do you actually need to override this?

**Review by stephentoub — APPROVED** [review]

**Comment by mgravell** [comment]
Woohoo! What's the merge politics/process for this repo? I have a "merge" button, but I don't think of runtime as "my" repo, so I don't want to abuse that button by clicking it if that isn't the process.

**Comment by stephentoub** [comment]
You can go ahead and merge if you'd like. Consider first removing the override I commented on if it's not actually needed.

**Comment by danmoseley** [comment]
Re process, if you get a sign off from someone on that team (as you did) and believe feedback is addressed and validation is green (or a known unrelated issue), you can merge yourself.

**Pull request merged**